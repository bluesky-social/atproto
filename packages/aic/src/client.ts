import { DidableKey } from '@adxp/crypto'
import axios from 'axios'
import { Document, hashAndFindDid } from './document'
import { Create, create, Operation, UpdateOperation } from './types'
import * as cbor from '@ipld/dag-cbor'
import * as uint8arrays from 'uint8arrays'
import { check, cidForData } from '@adxp/common'

export class AicClient {
  constructor(public url: string) {}

  async getDocument(did: string): Promise<Document> {
    const res = await axios.get(`${this.url}/${did}`)
    return res.data
  }

  async getOperationLog(did: string): Promise<Operation[]> {
    const res = await axios.get(`${this.url}/log/${did}`)
    return res.data.log
  }

  async createDid(
    signingKey: DidableKey,
    recoveryKey: string,
    username: string,
    service: string,
  ): Promise<string> {
    const op: Partial<Create> = {
      type: 'create',
      signingKey: signingKey.did(),
      recoveryKey,
      username,
      service,
      prev: null,
    }
    const signedOp = await this.signOperation(op, signingKey)
    if (!check.is(signedOp, create)) {
      throw new Error('Not a valid create operation')
    }
    const did = await hashAndFindDid(signedOp)
    await axios.post(`${this.url}/${did}`, signedOp)
    return did
  }

  async makeUpdate(
    did: string,
    op: Partial<UpdateOperation>,
    signingKey: DidableKey,
  ) {
    const log = await this.getOperationLog(did)
    if (log.length === 0) {
      throw new Error(`Could not make update: DID does not exist: ${did}`)
    }
    const prev = await cidForData(log[log.length - 1])
    const signedOp = this.signOperation(
      {
        ...op,
        prev: prev.toString(),
      },
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, signedOp)
  }

  async rotateSigningKey(did: string, newKey: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'rotate_signing_key', key: newKey },
      signingKey,
    )
  }

  async rotateRecoveryKey(did: string, newKey: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'rotate_recovery_key', key: newKey },
      signingKey,
    )
  }

  async udpateUsername(did: string, username: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'update_username', username },
      signingKey,
    )
  }

  async updateService(did: string, service: string, signingKey: DidableKey) {
    return this.makeUpdate(did, { type: 'update_service', service }, signingKey)
  }
}

export default AicClient
