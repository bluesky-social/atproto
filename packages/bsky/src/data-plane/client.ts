import assert from 'node:assert'
import { randomInt } from 'node:crypto'
import * as ui8 from 'uint8arrays'
import {
  Code,
  ConnectError,
  PromiseClient,
  createPromiseClient,
  makeAnyClient,
} from '@connectrpc/connect'
import { createGrpcTransport } from '@connectrpc/connect-node'
import { getDidKeyFromMultibase } from '@atproto/identity'
import { Service } from '../proto/bsky_connect'

export type DataPlaneClient = PromiseClient<typeof Service>
type BaseClient = { lib: DataPlaneClient; url: URL }
type HttpVersion = '1.1' | '2'
const MAX_RETRIES = 3

export const createDataPlaneClient = (
  baseUrls: string[],
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
) => {
  const clients = baseUrls.map((baseUrl) => createBaseClient(baseUrl, opts))
  assert(clients.length > 0, 'no clients available')
  return makeAnyClient(Service, (method) => {
    return async (...args) => {
      let tries = 0
      let error: unknown
      let remainingClients = clients
      while (tries < MAX_RETRIES) {
        const client = randomElement(remainingClients)
        assert(client, 'no clients available')
        try {
          return await client.lib[method.localName](...args)
        } catch (err) {
          if (
            err instanceof ConnectError &&
            (err.code === Code.Unavailable || err.code === Code.Aborted)
          ) {
            tries++
            error = err
            remainingClients = getRemainingClients(remainingClients, client)
          } else {
            throw err
          }
        }
      }
      assert(error)
      throw error
    }
  }) as DataPlaneClient
}

export { Code }

export const isDataplaneError = (
  err: unknown,
  code?: Code,
): err is ConnectError => {
  if (err instanceof ConnectError) {
    return !code || err.code === code
  }
  return false
}

const createBaseClient = (
  baseUrl: string,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
): BaseClient => {
  const { httpVersion = '2', rejectUnauthorized = true } = opts
  const transport = createGrpcTransport({
    baseUrl,
    httpVersion,
    acceptCompression: [],
    nodeOptions: { rejectUnauthorized },
  })
  return {
    lib: createPromiseClient(Service, transport),
    url: new URL(baseUrl),
  }
}

const getRemainingClients = (clients: BaseClient[], lastClient: BaseClient) => {
  if (clients.length < 2) return clients // no clients to choose from
  if (lastClient.url.port) {
    // if the last client had a port, we attempt to exclude its whole host.
    const maybeRemaining = clients.filter(
      (c) => c.url.hostname !== lastClient.url.hostname,
    )
    if (maybeRemaining.length) {
      return maybeRemaining
    }
  }
  return clients.filter((c) => c !== lastClient)
}

const randomElement = <T>(arr: T[]): T | undefined => {
  if (arr.length === 0) return
  return arr[randomInt(arr.length)]
}

export const unpackIdentityServices = (servicesBytes: Uint8Array) => {
  const servicesStr = ui8.toString(servicesBytes, 'utf8')
  if (!servicesStr) return {}
  return JSON.parse(servicesStr) as UnpackedServices
}

export const unpackIdentityKeys = (keysBytes: Uint8Array) => {
  const keysStr = ui8.toString(keysBytes, 'utf8')
  if (!keysStr) return {}
  return JSON.parse(keysStr) as UnpackedKeys
}

export const getServiceEndpoint = (
  services: UnpackedServices,
  opts: { id: string; type: string },
) => {
  const endpoint =
    services[opts.id] &&
    services[opts.id].Type === opts.type &&
    validateUrl(services[opts.id].URL)
  return endpoint || undefined
}

export const getKeyAsDidKey = (keys: UnpackedKeys, opts: { id: string }) => {
  const key =
    keys[opts.id] &&
    getDidKeyFromMultibase({
      type: keys[opts.id].Type,
      publicKeyMultibase: keys[opts.id].PublicKeyMultibase,
    })
  return key || undefined
}

type UnpackedServices = Record<string, { Type: string; URL: string }>

type UnpackedKeys = Record<string, { Type: string; PublicKeyMultibase: string }>

const validateUrl = (urlStr: string): string | undefined => {
  let url
  try {
    url = new URL(urlStr)
  } catch {
    return undefined
  }
  if (!['http:', 'https:'].includes(url.protocol)) {
    return undefined
  } else if (!url.hostname) {
    return undefined
  } else {
    return urlStr
  }
}
