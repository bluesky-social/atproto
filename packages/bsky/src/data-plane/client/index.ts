import assert from 'node:assert'
import { randomInt } from 'node:crypto'
import {
  Code,
  ConnectError,
  PromiseClient,
  createPromiseClient,
  makeAnyClient,
} from '@connectrpc/connect'
import { createGrpcTransport } from '@connectrpc/connect-node'
import { Service } from '../../proto/bsky_connect'
import { HostList } from './hosts'

export * from './hosts'
export * from './util'

export type DataPlaneClient = PromiseClient<typeof Service>
type HttpVersion = '1.1' | '2'
const MAX_RETRIES = 3

export const createDataPlaneClient = (
  hostList: HostList,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
) => {
  const clients = new DataPlaneClients(hostList, opts)
  return makeAnyClient(Service, (method) => {
    return async (...args) => {
      let tries = 0
      let error: unknown
      let remainingClients = clients.get()
      while (tries < MAX_RETRIES) {
        const client = randomElement(remainingClients)
        assert(client, 'no clients available')
        try {
          return await client[method.localName](...args)
        } catch (err) {
          if (
            err instanceof ConnectError &&
            (err.code === Code.Unavailable || err.code === Code.Aborted)
          ) {
            tries++
            error = err
            remainingClients = getRemainingClients(remainingClients, client)
          } else {
            throw err
          }
        }
      }
      assert(error)
      throw error
    }
  }) as DataPlaneClient
}

export { Code }

/**
 * Uses a reactive HostList in order to maintain a pool of DataPlaneClients.
 * Each DataPlaneClient is cached per host so that it maintains connections
 * and other internal state when the underlying HostList is updated.
 */
class DataPlaneClients {
  private clients: DataPlaneClient[] = []
  private clientsByHost = new Map<string, DataPlaneClient>()

  constructor(
    private hostList: HostList,
    private clientOpts: {
      httpVersion?: HttpVersion
      rejectUnauthorized?: boolean
    },
  ) {
    this.refresh()
    this.hostList.onUpdate(() => this.refresh())
  }

  get(): readonly DataPlaneClient[] {
    return this.clients
  }

  private refresh() {
    this.clients = []
    for (const host of this.hostList.get()) {
      let client = this.clientsByHost.get(host)
      if (!client) {
        client = this.createClient(host)
        this.clientsByHost.set(host, client)
      }
      this.clients.push(client)
    }
  }

  private createClient(host: string) {
    return createBaseClient(host, this.clientOpts)
  }
}

const createBaseClient = (
  baseUrl: string,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
): DataPlaneClient => {
  const { httpVersion = '2', rejectUnauthorized = true } = opts
  const transport = createGrpcTransport({
    baseUrl,
    httpVersion,
    acceptCompression: [],
    nodeOptions: { rejectUnauthorized },
  })
  return createPromiseClient(Service, transport)
}

const getRemainingClients = (
  clients: readonly DataPlaneClient[],
  lastClient: DataPlaneClient,
) => {
  if (clients.length < 2) return clients // no clients to choose from
  return clients.filter((c) => c !== lastClient)
}

const randomElement = <T>(arr: readonly T[]): T | undefined => {
  if (arr.length === 0) return
  return arr[randomInt(arr.length)]
}
