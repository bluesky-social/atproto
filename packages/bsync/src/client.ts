import {
  Interceptor,
  PromiseClient,
  createPromiseClient,
} from '@connectrpc/connect'
import {
  ConnectTransportOptions,
  createConnectTransport,
} from '@connectrpc/connect-node'
import { Service } from './proto/bsync_connect'

export type BsyncClient = PromiseClient<typeof Service>

export const createClient = (opts: ConnectTransportOptions): BsyncClient => {
  const transport = createConnectTransport(opts)
  return createPromiseClient(Service, transport)
}

export const authWithApiKey =
  (apiKey: string): Interceptor =>
  (next) =>
  (req) => {
    req.header.set('authorization', `Bearer ${apiKey}`)
    return next(req)
  }
