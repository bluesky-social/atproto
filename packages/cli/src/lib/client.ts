import path from 'path'
import {
  IpldStore,
  MicroblogClient,
  MicroblogDelegator,
  MicroblogFull,
  Repo,
} from '@atproto/common'
import * as config from '../lib/config'

export const loadClient = async (
  repoPath: string,
): Promise<MicroblogClient> => {
  const cfg = await config.loadCfg(repoPath)
  if (cfg.account.delegator) {
    return loadDelegate(cfg)
  } else {
    return loadFull(cfg, repoPath)
  }
}

export const loadFull = async (
  cfg: config.Config,
  repoPath: string,
): Promise<MicroblogFull> => {
  const { account, authStore, root } = cfg
  const blockstore = IpldStore.createPersistent(
    path.join(repoPath, 'blockstore'),
  )
  let repo: Repo
  if (!root) {
    repo = await Repo.create(blockstore, await authStore.did(), authStore)
  } else {
    repo = await Repo.load(blockstore, root, authStore)
  }
  return new MicroblogFull(repo, `http://${account.server}`, {
    onPush: async (newRoot) => {
      await config.writeRoot(repoPath, newRoot)
    },
  })
}

export const loadDelegate = async (
  cfg: config.Config,
): Promise<MicroblogDelegator> => {
  const { account, authStore } = cfg
  return new MicroblogDelegator(
    `http://${account.server}`,
    await authStore.did(),
    authStore,
  )
}
