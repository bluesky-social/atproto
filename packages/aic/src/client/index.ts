import axios from 'axios'
import { CID } from 'multiformats/cid'
import { DidableKey } from '@adxp/crypto'
import { check, cidForData } from '@adxp/common'
import * as operations from '../lib/operations'
import * as t from '../lib/types'

export class AicClient {
  constructor(public url: string) {}

  async getDocument(did: string): Promise<t.DocumentData> {
    const res = await axios.get(`${this.url}/${did}`)
    return res.data
  }

  async getOperationLog(did: string): Promise<t.Operation[]> {
    const res = await axios.get(`${this.url}/log/${did}`)
    return res.data.log
  }

  async createDid(
    signingKey: DidableKey,
    recoveryKey: string,
    username: string,
    service: string,
  ): Promise<string> {
    const op = await operations.create(
      signingKey,
      recoveryKey,
      username,
      service,
    )
    if (!check.is(op, t.def.createOp)) {
      throw new Error('Not a valid create operation')
    }
    const did = await operations.didForCreateOp(op)
    await axios.post(`${this.url}/${did}`, op)
    return did
  }

  async getPrev(did): Promise<CID> {
    const log = await this.getOperationLog(did)
    if (log.length === 0) {
      throw new Error(`Could not make update: DID does not exist: ${did}`)
    }
    return cidForData(log[log.length - 1])
  }

  async rotateSigningKey(did: string, newKey: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.rotateSigningKey(
      newKey,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }

  async rotateRecoveryKey(did: string, newKey: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.rotateRecoveryKey(
      newKey,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }

  async updateUsername(did: string, username: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.updateUsername(
      username,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }

  async updateService(did: string, service: string, signingKey: DidableKey) {
    const prev = await this.getPrev(did)
    const op = await operations.updateService(
      service,
      prev.toString(),
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, op)
  }
}

export default AicClient
