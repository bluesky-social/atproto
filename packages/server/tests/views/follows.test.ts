import AdxApi, { ServiceClient as AdxServiceClient } from '@adxp/api'
import { runTestServer, forSnapshot, CloseFn, constantDate } from '../_util'
import { SeedClient } from '../seeds/client'
import followsSeed from '../seeds/follows'

describe('pds follow views', () => {
  let client: AdxServiceClient
  let close: CloseFn
  let sc: SeedClient

  // account dids, for convenience
  let alice: string

  beforeAll(async () => {
    const server = await runTestServer({
      dbPostgresSchema: 'views_follows',
    })
    close = server.close
    client = AdxApi.service(server.url)
    sc = new SeedClient(client)
    await followsSeed(sc)
    alice = sc.dids.alice
  })

  afterAll(async () => {
    await close()
  })

  const getCursors = (items: { createdAt?: string }[]) =>
    items.map((item) => item.createdAt ?? constantDate)

  const getSortedCursors = (items: { createdAt?: string }[]) =>
    getCursors(items).sort((a, b) => tstamp(b) - tstamp(a))

  const tstamp = (x: string) => new Date(x).getTime()

  it('fetches followers', async () => {
    const aliceFollowers = await client.todo.social.getUserFollowers({
      user: sc.dids.alice,
    })

    expect(forSnapshot(aliceFollowers.data)).toMatchSnapshot()
    expect(getCursors(aliceFollowers.data.followers)).toEqual(
      getSortedCursors(aliceFollowers.data.followers),
    )

    const bobFollowers = await client.todo.social.getUserFollowers({
      user: sc.dids.bob,
    })

    expect(forSnapshot(bobFollowers.data)).toMatchSnapshot()
    expect(getCursors(bobFollowers.data.followers)).toEqual(
      getSortedCursors(bobFollowers.data.followers),
    )

    const carolFollowers = await client.todo.social.getUserFollowers({
      user: sc.dids.carol,
    })

    expect(forSnapshot(carolFollowers.data)).toMatchSnapshot()
    expect(getCursors(carolFollowers.data.followers)).toEqual(
      getSortedCursors(carolFollowers.data.followers),
    )

    const danFollowers = await client.todo.social.getUserFollowers({
      user: sc.dids.dan,
    })

    expect(forSnapshot(danFollowers.data)).toMatchSnapshot()
    expect(getCursors(danFollowers.data.followers)).toEqual(
      getSortedCursors(danFollowers.data.followers),
    )

    const eveFollowers = await client.todo.social.getUserFollowers({
      user: sc.dids.eve,
    })

    expect(forSnapshot(eveFollowers.data)).toMatchSnapshot()
    expect(getCursors(eveFollowers.data.followers)).toEqual(
      getSortedCursors(eveFollowers.data.followers),
    )
  })

  it('fetches followers by username', async () => {
    const byDid = await client.todo.social.getUserFollowers({
      user: sc.dids.alice,
    })
    const byUsername = await client.todo.social.getUserFollowers({
      user: sc.accounts[alice].username,
    })
    expect(byUsername.data).toEqual(byDid.data)
  })

  it('paginates followers', async () => {
    const full = await client.todo.social.getUserFollowers({
      user: sc.dids.alice,
    })

    expect(full.data.followers.length).toEqual(4)

    const paginated = await client.todo.social.getUserFollowers({
      user: sc.dids.alice,
      before: full.data.followers[0].createdAt,
      limit: 2,
    })

    expect(paginated.data.followers).toEqual(full.data.followers.slice(1, 3))
  })

  it('fetches follows', async () => {
    const aliceFollowers = await client.todo.social.getUserFollows({
      user: sc.dids.alice,
    })

    expect(forSnapshot(aliceFollowers.data)).toMatchSnapshot()
    expect(getCursors(aliceFollowers.data.follows)).toEqual(
      getSortedCursors(aliceFollowers.data.follows),
    )

    const bobFollowers = await client.todo.social.getUserFollows({
      user: sc.dids.bob,
    })

    expect(forSnapshot(bobFollowers.data)).toMatchSnapshot()
    expect(getCursors(bobFollowers.data.follows)).toEqual(
      getSortedCursors(bobFollowers.data.follows),
    )

    const carolFollowers = await client.todo.social.getUserFollows({
      user: sc.dids.carol,
    })

    expect(forSnapshot(carolFollowers.data)).toMatchSnapshot()
    expect(getCursors(carolFollowers.data.follows)).toEqual(
      getSortedCursors(carolFollowers.data.follows),
    )

    const danFollowers = await client.todo.social.getUserFollows({
      user: sc.dids.dan,
    })

    expect(forSnapshot(danFollowers.data)).toMatchSnapshot()
    expect(getCursors(danFollowers.data.follows)).toEqual(
      getSortedCursors(danFollowers.data.follows),
    )

    const eveFollowers = await client.todo.social.getUserFollows({
      user: sc.dids.eve,
    })

    expect(forSnapshot(eveFollowers.data)).toMatchSnapshot()
    expect(getCursors(eveFollowers.data.follows)).toEqual(
      getSortedCursors(eveFollowers.data.follows),
    )
  })

  it('fetches follows by username', async () => {
    const byDid = await client.todo.social.getUserFollows({
      user: sc.dids.alice,
    })
    const byUsername = await client.todo.social.getUserFollows({
      user: sc.accounts[alice].username,
    })
    expect(byUsername.data).toEqual(byDid.data)
  })

  it('paginates follows', async () => {
    const full = await client.todo.social.getUserFollows({
      user: sc.dids.alice,
    })

    expect(full.data.follows.length).toEqual(4)

    const paginated = await client.todo.social.getUserFollows({
      user: sc.dids.alice,
      before: full.data.follows[0].createdAt,
      limit: 2,
    })

    expect(paginated.data.follows).toEqual(full.data.follows.slice(1, 3))
  })
})
