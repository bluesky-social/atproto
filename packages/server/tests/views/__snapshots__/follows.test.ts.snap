// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pds follow views fetches followers 1`] = `
Object {
  "followers": Array [
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(1)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "eve.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(2)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "dan.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(3)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "bob.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(4)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "carol.test",
    },
  ],
  "subject": Object {
    "did": "user(0)",
    "name": "alice.test",
  },
}
`;

exports[`pds follow views fetches followers 2`] = `
Object {
  "followers": Array [
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(1)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "dan.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(2)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "alice.test",
    },
  ],
  "subject": Object {
    "did": "user(0)",
    "name": "bob.test",
  },
}
`;

exports[`pds follow views fetches followers 3`] = `
Object {
  "followers": Array [
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(1)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "eve.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(2)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "bob.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(3)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "alice.test",
    },
  ],
  "subject": Object {
    "did": "user(0)",
    "name": "carol.test",
  },
}
`;

exports[`pds follow views fetches followers 4`] = `
Object {
  "followers": Array [
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(1)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "alice.test",
    },
  ],
  "subject": Object {
    "did": "user(0)",
    "name": "dan.test",
  },
}
`;

exports[`pds follow views fetches followers 5`] = `
Object {
  "followers": Array [
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(1)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "dan.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(2)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "alice.test",
    },
  ],
  "subject": Object {
    "did": "user(0)",
    "name": "eve.test",
  },
}
`;

exports[`pds follow views fetches follows 1`] = `
Object {
  "follows": Array [
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(1)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "eve.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(2)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "dan.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(3)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "carol.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(4)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "bob.test",
    },
  ],
  "subject": Object {
    "did": "user(0)",
    "name": "alice.test",
  },
}
`;

exports[`pds follow views fetches follows 2`] = `
Object {
  "follows": Array [
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(1)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "carol.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(2)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "alice.test",
    },
  ],
  "subject": Object {
    "did": "user(0)",
    "name": "bob.test",
  },
}
`;

exports[`pds follow views fetches follows 3`] = `
Object {
  "follows": Array [
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(1)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "alice.test",
    },
  ],
  "subject": Object {
    "did": "user(0)",
    "name": "carol.test",
  },
}
`;

exports[`pds follow views fetches follows 4`] = `
Object {
  "follows": Array [
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(1)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "eve.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(2)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "bob.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(3)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "alice.test",
    },
  ],
  "subject": Object {
    "did": "user(0)",
    "name": "dan.test",
  },
}
`;

exports[`pds follow views fetches follows 5`] = `
Object {
  "follows": Array [
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(1)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "carol.test",
    },
    Object {
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(2)",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "alice.test",
    },
  ],
  "subject": Object {
    "did": "user(0)",
    "name": "eve.test",
  },
}
`;
